src/mvba/abba/mod.rs

```diff
@@ -1,56 +1,567 @@
+                            .filter(|(_, a)| a.value == MainVoteValue::Abstain)
+                            .count();
+
+                        let (value, justification) = if zero_count > 0 {
+                            // if there is a main-vote for 0,
+                            let sig = match &main_votes.iter().last().unwrap().1.justification {
+                                MainVoteJustification::NoAbstainJustification(sig) => sig,
+                                _ => {
+                                    return Err(Error::Generic(
+                                        "protocol violated, invalid main-vote justification"
+                                            .to_string(),
+                                    ))
+                                }
+                            };
+                            // hard pre-vote for 0
+                            (
+                                PreVoteValue::Zero,
+                                PreVoteJustification::HardJustification(sig.clone()),
+                            )
+                        } else if one_count > 1 {
+                            // if there is a main-vote for 1,
+                            let sig = match &main_votes.iter().last().unwrap().1.justification {
+                                MainVoteJustification::NoAbstainJustification(sig) => sig,
+                                _ => {
+                                    return Err(Error::Generic(
+                                        "protocol violated, invalid main-vote justification"
+                                            .to_string(),
+                                    ))
+                                }
+                            };
+                            // hard pre-vote for 1
+                            (
+                                PreVoteValue::One,
+                                PreVoteJustification::HardJustification(sig.clone()),
+                            )
+                        } else if abstain_count == self.threshold() {
+                            // if all main-votes are abstain,
+                            let sig = match &main_votes.iter().last().unwrap().1.justification {
+                                MainVoteJustification::AbstainJustification(_, _) => {
+                                    let sig_share: HashMap<&NodeId, &SignatureShare> =
+                                        main_votes.iter().map(|(n, a)| (n, &a.sig_share)).collect();
+                                    self.pub_key_set.combine_signatures(sig_share)?
+                                }
+                                _ => {
+                                    return Err(Error::Generic(
+                                        "protocol violated, invalid main-vote justification"
+                                            .to_string(),
+                                    ))
+                                }
+                            };
```

Comment:
This match seems to be unnecessary, it seems we can just aggregate the signatures from the abstains:
```suggestion
                            let sig_share: HashMap<&NodeId, &SignatureShare> =
                                main_votes.iter().map(|(n, a)| (n, &a.sig_share)).collect();
                            let sig = self.pub_key_set.combine_signatures(sig_share)?
```